cloudinit/net/__init__.py

```diff
@@ -971,18 +972,38 @@ def get_ib_hwaddrs_by_interface():
     return ret
 
 
-def has_url_connectivity(url):
-    """Return true when the instance has access to the provided URL
+def has_url_connectivity(
+    url_data: Union[MutableMapping[str, Any], str]
+) -> bool:
+    """Return true when the instance has access to the provided URL.
 
     Logs a warning if url is not the expected format.
+
+    url_data is either the URL string to check, or a dictionary of kwargs
+    to send to readurl. E.g.:
+
+    has_url_connectivity("http://example.invalid")
```

Comment:
Just for a clean/simple function signature and expectations, I'd vote to drop the "str" type option if we can. Less areas where call-sites could misinterpret what needs sending..